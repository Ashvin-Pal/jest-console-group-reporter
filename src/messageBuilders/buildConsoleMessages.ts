import { buildConsoleMessage } from "./buildConsoleMessage";
import type { ConsoleMessagesMap, DisplayOptions, LogEntry, Options } from "@/types";

const orderByCount = (
  [, { count: aCount }]: [string, LogEntry],
  [, { count: bCount }]: [string, LogEntry]
): number => {
  if (aCount > bCount) return -1;
  if (bCount > aCount) return 1;

  return 0;
};

export const buildConsoleMessages = ({
  consoleLevels,
  consoleMap,
  displayOptions,
  useGitHubActions,
}: {
  consoleMap: ConsoleMessagesMap;
  consoleLevels: Options["consoleLevels"];
  useGitHubActions: Options["useGitHubActions"];
  displayOptions: DisplayOptions;
}): string[] => {
  if (consoleMap.size === 0) {
    return [];
  }

  return consoleLevels.reduce((acc: string[], type) => {
    const consoleType = consoleMap.get(type);
    if (consoleType) {
      const consoleTypeEntries = Array.from(consoleType.entries()).sort(orderByCount);
      const consoleMessage = consoleTypeEntries.map(([message, logEntry]) =>
        buildConsoleMessage({ message, type, ...logEntry, displayOptions, useGitHubActions })
      );
      acc.push(...consoleMessage);
    }
    return acc;
  }, []);
};
