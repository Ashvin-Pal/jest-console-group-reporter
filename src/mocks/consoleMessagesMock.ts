import type { ConsoleMessage } from "@/types";

export const consoleMessagesMock: ConsoleMessage[] = [
  {
    type: "error",
    message: "Network connection lost.",
    origin: "Network Module",
  },
  {
    type: "error",
    message: "Network connection lost.",
    origin: "Network Module",
  },
  {
    type: "error",
    message: "There was an error processing the data.",
    origin: "Server",
  },
  {
    type: "warn",
    message: "Server timeout.",
    origin: "Server",
  },
  {
    type: "warn",
    message: "Resource not found.",
    origin: "Resource Loader",
  },
  {
    type: "info",
    message: "User profile updated.",
    origin: "User Profile",
  },
  {
    type: "info",
    message:
      "This is an exceptionally lengthy info message. It ensures that extremely long messages are truncated when they exceed 150 characters in test summaries.",
    origin: "User Profile",
  },
  {
    type: "log",
    message: "This is a sample log message.",
    origin: "App",
  },
];
