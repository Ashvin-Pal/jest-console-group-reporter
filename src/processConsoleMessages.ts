import type { TestResult, Config } from "@jest/reporters";
import type { ConsoleMessagesMap, Options } from "@/types";
import { getGroupName } from "./getGroupName";
import { shouldFilterMessage } from "./shouldFilterMessage";
import { mergeLogEntries, createLogEntry } from "@/logEntry";

interface ProcessConsoleMessagesOptions
  extends Pick<Options, "filters" | "groups">,
    Pick<TestResult, "testFilePath" | "console"> {
  config: Config.ProjectConfig;
}

export function processConsoleMessages({
  filters,
  groups,
  testFilePath,
  console: consoleMessages = [],
  config,
}: ProcessConsoleMessagesOptions): {
  consoleMessagesMap: ConsoleMessagesMap;
  filteredCount: number;
} {
  const consoleMessagesMap: ConsoleMessagesMap = new Map();
  let filteredCount = 0;

  for (const consoleMessage of consoleMessages) {
    if (shouldFilterMessage(filters, consoleMessage)) {
      filteredCount += 1;
      continue;
    }

    const { groupName, isCustomGroup } = getGroupName(consoleMessage, groups);

    let logEntryMap = consoleMessagesMap.get(consoleMessage.type);
    if (!logEntryMap) {
      logEntryMap = new Map();
      consoleMessagesMap.set(consoleMessage.type, logEntryMap);
    }

    const logEntry = createLogEntry({ isCustomGroup, testFilePath, ...consoleMessage, config });

    const existingEntry = logEntryMap.get(groupName);
    logEntryMap.set(groupName, existingEntry ? mergeLogEntries(existingEntry, logEntry) : logEntry);
  }

  return { consoleMessagesMap, filteredCount };
}
