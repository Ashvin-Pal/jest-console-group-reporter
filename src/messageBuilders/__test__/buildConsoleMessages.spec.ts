import type { Config } from "@jest/reporters";
import { buildConsoleMessages } from "../buildConsoleMessages";
import { consoleMessagesMock } from "../../mocks";
import { processConsoleMessages } from "../../processConsoleMessages";

const testFilePath = "test.js";
const config = { projectRoot: "./" } as unknown as Config.ProjectConfig;

describe("buildConsoleMessages", () => {
  test("Build [summary] console messages correctly when there are no group configurations", () => {
    const { consoleMessagesMap } = processConsoleMessages({
      filters: [],
      groups: [],
      testFilePath,
      // @ts-expect-error Issues with jest type
      console: consoleMessagesMock,
      config,
    });

    expect(
      buildConsoleMessages({
        consoleMap: consoleMessagesMap,
        consoleLevels: ["error", "warn", "log", "info", "log"],
        useGitHubActions: false,
        displayOptions: {
          enabled: true,
          filePaths: false,
          reportType: "summary",
        },
      })
    ).toMatchSnapshot();
  });

  test("Build [summary] console messages correctly when there are group configurations", () => {
    const { consoleMessagesMap } = processConsoleMessages({
      filters: [],
      groups: [
        {
          name: "Error group",
          match: ({ type }): boolean => type === "error",
        },
      ],
      testFilePath,
      // @ts-expect-error Issues with jest type
      console: consoleMessagesMock,
      config,
    });

    expect(
      buildConsoleMessages({
        consoleMap: consoleMessagesMap,
        consoleLevels: ["error", "warn", "log", "info", "debug"],
        useGitHubActions: false,
        displayOptions: {
          enabled: true,
          filePaths: false,
          reportType: "summary",
        },
      })
    ).toMatchSnapshot();
  });

  test("Build [detailed] console messages correctly when there are group configurations", () => {
    const { consoleMessagesMap } = processConsoleMessages({
      filters: [],
      groups: [],
      testFilePath,
      // @ts-expect-error Issues with jest type
      console: consoleMessagesMock,
      config,
    });

    expect(
      buildConsoleMessages({
        consoleMap: consoleMessagesMap,
        consoleLevels: ["error", "warn", "log", "info", "log"],
        useGitHubActions: false,
        displayOptions: {
          enabled: true,
          filePaths: true,
          reportType: "detailed",
        },
      })
    ).toMatchSnapshot();
  });

  test("Build [detailed] console messages correctly when there are group configurations", () => {
    const { consoleMessagesMap } = processConsoleMessages({
      filters: [],
      groups: [
        {
          name: "Error group",
          match: ({ type }): boolean => type === "error",
        },
      ],
      testFilePath,
      // @ts-expect-error Issues with jest type
      console: consoleMessagesMock,
      config,
    });

    expect(
      buildConsoleMessages({
        consoleMap: consoleMessagesMap,
        consoleLevels: ["error", "warn", "log", "info", "debug"],
        useGitHubActions: false,
        displayOptions: {
          enabled: true,
          filePaths: true,
          reportType: "detailed",
        },
      })
    ).toMatchSnapshot();
  });

  test("Build [detailed] console messages correctly when using the useGithubActions configuration", () => {
    const { consoleMessagesMap } = processConsoleMessages({
      filters: [],
      groups: [
        {
          name: "Error group",
          match: ({ type }): boolean => type === "error",
        },
      ],
      testFilePath,
      // @ts-expect-error Issues with jest type
      console: consoleMessagesMock,
      config,
    });

    expect(
      buildConsoleMessages({
        consoleMap: consoleMessagesMap,
        consoleLevels: ["error", "warn", "log", "info", "debug"],
        useGitHubActions: true,
        displayOptions: {
          enabled: true,
          filePaths: true,
          reportType: "detailed",
        },
      })
    ).toMatchSnapshot();
  });
});
