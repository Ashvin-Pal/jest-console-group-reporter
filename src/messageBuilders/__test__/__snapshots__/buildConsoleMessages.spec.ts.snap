// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildConsoleMessages Build [detailed] console messages correctly when there are group configurations 1`] = `
[
  "    [Bold-Red] ERROR   [Bold-White] 2    
[Red]       Network connection lost.
[Gray] Network Module

    Test File Path:
      1. test.js
",
  "    [Bold-Red] ERROR   [Bold-White] 1    
[Red]       There was an error processing the data.
[Gray] Server

    Test File Path:
      1. test.js
",
  "    [Bold-Yellow] WARN    [Bold-White] 1    
[Yellow]       Server timeout.
[Gray] Server

    Test File Path:
      1. test.js
",
  "    [Bold-Yellow] WARN    [Bold-White] 1    
[Yellow]       Resource not found.
[Gray] Resource Loader

    Test File Path:
      1. test.js
",
  "    [Bold-Cyan] LOG     [Bold-White] 1    
[Cyan]       This is a sample log message.
[Gray] App

    Test File Path:
      1. test.js
",
  "    [Bold-Blue] INFO    [Bold-White] 1    
[Blue]       User profile updated.
[Gray] User Profile

    Test File Path:
      1. test.js
",
  "    [Bold-Blue] INFO    [Bold-White] 1    
[Blue]       This is an exceptionally lengthy info message. It ensures that extremely long messages are truncated when they exceed 150 characters in test summaries.
[Gray] User Profile

    Test File Path:
      1. test.js
",
  "    [Bold-Cyan] LOG     [Bold-White] 1    
[Cyan]       This is a sample log message.
[Gray] App

    Test File Path:
      1. test.js
",
]
`;

exports[`buildConsoleMessages Build [detailed] console messages correctly when there are group configurations 2`] = `
[
  "    [Bold-Red] ERROR   [Bold-White] 3    
[Red]       Error group
[Gray] Network Module

    Test File Path:
      1. test.js
",
  "    [Bold-Yellow] WARN    [Bold-White] 1    
[Yellow]       Server timeout.
[Gray] Server

    Test File Path:
      1. test.js
",
  "    [Bold-Yellow] WARN    [Bold-White] 1    
[Yellow]       Resource not found.
[Gray] Resource Loader

    Test File Path:
      1. test.js
",
  "    [Bold-Cyan] LOG     [Bold-White] 1    
[Cyan]       This is a sample log message.
[Gray] App

    Test File Path:
      1. test.js
",
  "    [Bold-Blue] INFO    [Bold-White] 1    
[Blue]       User profile updated.
[Gray] User Profile

    Test File Path:
      1. test.js
",
  "    [Bold-Blue] INFO    [Bold-White] 1    
[Blue]       This is an exceptionally lengthy info message. It ensures that extremely long messages are truncated when they exceed 150 characters in test summaries.
[Gray] User Profile

    Test File Path:
      1. test.js
",
]
`;

exports[`buildConsoleMessages Build [detailed] console messages correctly when using the useGithubActions configuration 1`] = `
[
  "::group::    [Bold-Red] ERROR   [Bold-White] 3    [Red] Error group

[Red]       Error group
[Gray] Network Module

    Test File Path:
      1. test.js

::endgroup::",
  "::group::    [Bold-Yellow] WARN    [Bold-White] 1    [Yellow] Server timeout.

[Yellow]       Server timeout.
[Gray] Server

    Test File Path:
      1. test.js

::endgroup::",
  "::group::    [Bold-Yellow] WARN    [Bold-White] 1    [Yellow] Resource not found.

[Yellow]       Resource not found.
[Gray] Resource Loader

    Test File Path:
      1. test.js

::endgroup::",
  "::group::    [Bold-Cyan] LOG     [Bold-White] 1    [Cyan] This is a sample log message.

[Cyan]       This is a sample log message.
[Gray] App

    Test File Path:
      1. test.js

::endgroup::",
  "::group::    [Bold-Blue] INFO    [Bold-White] 1    [Blue] User profile updated.

[Blue]       User profile updated.
[Gray] User Profile

    Test File Path:
      1. test.js

::endgroup::",
  "::group::    [Bold-Blue] INFO    [Bold-White] 1    [Blue] This is an exceptionally lengthy info message. It ensures that extremely long messages are truncated when they exceed 150 characters in test summaries...

[Blue]       This is an exceptionally lengthy info message. It ensures that extremely long messages are truncated when they exceed 150 characters in test summaries.
[Gray] User Profile

    Test File Path:
      1. test.js

::endgroup::",
]
`;

exports[`buildConsoleMessages Build [summary] console messages correctly when there are group configurations 1`] = `
[
  "    [Bold-Red] ERROR   [Bold-White] 3    [Red] Error group",
  "    [Bold-Yellow] WARN    [Bold-White] 1    [Yellow] Server timeout.",
  "    [Bold-Yellow] WARN    [Bold-White] 1    [Yellow] Resource not found.",
  "    [Bold-Cyan] LOG     [Bold-White] 1    [Cyan] This is a sample log message.",
  "    [Bold-Blue] INFO    [Bold-White] 1    [Blue] User profile updated.",
  "    [Bold-Blue] INFO    [Bold-White] 1    [Blue] This is an exceptionally lengthy info message. It ensures that extremely long messages are truncated when they exceed 150 characters in test summaries...",
]
`;

exports[`buildConsoleMessages Build [summary] console messages correctly when there are no group configurations 1`] = `
[
  "    [Bold-Red] ERROR   [Bold-White] 2    [Red] Network connection lost.",
  "    [Bold-Red] ERROR   [Bold-White] 1    [Red] There was an error processing the data.",
  "    [Bold-Yellow] WARN    [Bold-White] 1    [Yellow] Server timeout.",
  "    [Bold-Yellow] WARN    [Bold-White] 1    [Yellow] Resource not found.",
  "    [Bold-Cyan] LOG     [Bold-White] 1    [Cyan] This is a sample log message.",
  "    [Bold-Blue] INFO    [Bold-White] 1    [Blue] User profile updated.",
  "    [Bold-Blue] INFO    [Bold-White] 1    [Blue] This is an exceptionally lengthy info message. It ensures that extremely long messages are truncated when they exceed 150 characters in test summaries...",
  "    [Bold-Cyan] LOG     [Bold-White] 1    [Cyan] This is a sample log message.",
]
`;
